chia/wallet/nft_wallet/nft_wallet.py

```diff
@@ -930,3 +932,210 @@ async def set_nft_did(self, nft_coin_info: NFTCoinInfo, did_id: bytes, fee: uint
+        xch_change_ph: Optional[bytes32] = None,
+        new_innerpuzhash: Optional[bytes32] = None,
+        did_coin: Optional[Coin] = None,
+        did_lineage_parent: Optional[bytes32] = None,
+        fee: Optional[uint64] = uint64(0),
+    ):
+        # get DID Wallet
+        for wallet_id in self.wallet_state_manager.wallets:
+            wallet = self.wallet_state_manager.wallets[wallet_id]
+            if wallet.type() == WalletType.DECENTRALIZED_ID.value:
+                if self.get_did() == hexstr_to_bytes(wallet.get_my_DID()):
+                    did_wallet = wallet
+                    break
+        else:
+            raise ValueError("There is no DID associated with this NFT wallet")
+
+        assert did_wallet.did_info.current_inner is not None
+        assert did_wallet.did_info.origin_coin is not None
+        if max_num is None:
+            max_num = len(metadata_list)
+        assert isinstance(starting_num, int)
+        assert len(metadata_list) <= max_num + 1 - starting_num
+        if did_coin is None:
+            coins = await did_wallet.select_coins(uint64(1))
+            assert coins is not None
+            did_coin = coins.pop()
+
+        assert isinstance(fee, uint64)
+        total_amount = len(metadata_list) + fee
+        if xch_coins is None:
+            xch_coins = await self.standard_wallet.select_coins(uint64(total_amount))
+        assert len(xch_coins) > 0
+        spend_value = sum([coin.amount for coin in xch_coins])
+        change: uint64 = uint64(spend_value - total_amount)
+        xch_spends = []
+        if xch_change_ph is None:
+            xch_change_ph = await self.standard_wallet.get_new_puzzlehash()
+        xch_primaries = [
+            AmountWithPuzzlehash({"puzzlehash": xch_change_ph, "amount": change, "memos": [xch_change_ph]})
+        ]
+        for xch_coin in xch_coins:
+            puzzle: Program = await self.standard_wallet.puzzle_for_puzzle_hash(xch_coin.puzzle_hash)
+            solution: Program = self.standard_wallet.make_solution(
+                primaries=xch_primaries,
+                fee=fee,
+            )
+            xch_spends.append(CoinSpend(xch_coin, puzzle, solution))
+        xch_spend = await self.standard_wallet.sign_transaction(xch_spends)
+
+        innerpuz: Program = did_wallet.did_info.current_inner
+        # Quote message puzzle & solution
+        if new_innerpuzhash is None:
+            new_innerpuzhash = innerpuz.get_tree_hash()
+
+        # create primaries
+        primaries = [
+            AmountWithPuzzlehash(
+                {"puzzlehash": new_innerpuzhash, "amount": uint64(did_coin.amount), "memos": [new_innerpuzhash]}
+            )
+        ]
+
+        n = len(metadata_list)
+        amount = uint64(1)
+        zero_coin_spends = []
+        launcher_spends = []
+        launcher_ids = []
+        eve_spends = []
+        p2_inner_puzzle = await self.standard_wallet.get_new_puzzle()
+        ending_num = starting_num + n
```

Comment:
what's the difference between `max_num` and `ending_num`?

Reply:
max_num is equivalent to edition_total, ie the total number of NFTs that are going to be minted in this run. Then ending_num is the the highest value for edition_number that will be used in this chunk of NFTs getting put in a spend bundle. So we pass a metadata_list, with a starting_num, and we use this as the m value in the ephemeral puzzles from starting_num up to ending_num. So ending_num is the max m value we'll use in this run, and max_num is always fixed as the edition_total (the n in our m-of-n ephemeral puzzle)...bit confusing, sorry. Will add comments as requested above also.